File: go/internal/cbor/deterministic_test.go
package cbor

import (
	"io/ioutil"
	"os"
	"testing"
)

var (
	// Deterministically encoded unsigned integers.
	uint0          = []byte{0x00}
	uint10         = []byte{0x0A}
	uint23         = []byte{0x17}
	uint24         = []byte{0x18, 0x18}
	uint45         = []byte{0x18, 0x2D}
	uint255        = []byte{0x18, 0xFF}
	uint256        = []byte{0x19, 0x01, 0x00}
	uint5000       = []byte{0x19, 0x13, 0x88}
	uint65535      = []byte{0x19, 0xFF, 0xFF}
	uint65536      = []byte{0x1A, 0x00, 0x01, 0x00, 0x00}
	uint4294967    = []byte{0x1A, 0x00, 0x41, 0x89, 0x37}
	uint4294967295 = []byte{0x1A, 0xFF, 0xFF, 0xFF, 0xFF}
	uint4294967296 = []byte{0x1B, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00}

	// Long.MAX_VALUE aka 0x7F 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF.
	uint9223372036854775807 = []byte{0x1B, 0x7F, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}

	// Max CBOR supported value 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF.
	uint18446744073709551615 = []byte{0x1B, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF, 0xFF}

	uint64TestCasesAsBytes = [][]byte{uint0, uint10, uint23, uint24, uint45, uint255, uint256, uint5000, uint65535, uint65536, uint4294967, uint4294967295, uint4294967296, uint9223372036854775807, uint18446744073709551615}
	uint64Values           = []uint64{0, 10, 23, 24, 45, 255, 256, 5000, 65535, 65536, 4294967, 4294967295, 4294967296, 9223372036854775807, 18446744073709551615}
)

func TestUintDeterministic(t *testing.T) {
	for _, testCase := range uint64TestCasesAsBytes {
		if err := Deterministic(testCase); err != nil {
			t.Error("Deterministically encoded unsigned integers should not return error.")
		}
	}
}

func TestUintCborSequenceDeterministic(t *testing.T) {
	concatenatedTestCases := multiappend(uint64TestCasesAsBytes...)
	if err := Deterministic(concatenatedTestCases); err != nil {
		t.Error("Deterministically encoded unsigned integers should not return false for deterministicy.")
	}
}

// TestNotDeterministic tests that if the value of the unsigned integer is kept the same but there are
// additional non-necessary empty bytes added in-front of the actual value's bytes, it is no longer
// considered as deterministically encoded CBOR.
func TestUintNotDeterministic(t *testing.T) {
	// The additional information values for representing the number with 1, 2, 4 or 8 bytes.
	firstBytes := []byte{0x18, 0x19, 0x1A, 0x1B}

	for _, testCase := range uint64TestCasesAsBytes {
		for _, firstByte := range firstBytes {
			ainfo := convertToAdditionalInfo(firstByte)
			newLength := ainfo.getAdditionalInfoLength()

			// Cannot represent too big number with too little amount of bytes.
			if len(testCase) >= newLength {
				continue
			}

			nonDeterministicBytes := convertToNonDeterministicUintHelper(testCase, firstByte)
			valueOfOriginalByteArr := getUnsignedIntegerValue(testCase, convertToAdditionalInfo(testCase[0]))
			valueOfNonDeterministicByteArr := getUnsignedIntegerValue(nonDeterministicBytes, convertToAdditionalInfo(nonDeterministicBytes[0]))

			if valueOfOriginalByteArr != valueOfNonDeterministicByteArr {
				t.Error("valueOfOriginalByteArr and valueOfNonDeterministicByteArr should match.")
			}

			if err := Deterministic(nonDeterministicBytes); err == nil {
				t.Error("Non-deterministically encoded unsigned integers should not return true for deterministicy.")
			}
		}
	}
}

func TestEmptyByteArrayIsDeterministic(t *testing.T) {
	if err := Deterministic([]byte{}); err != nil {
		t.Error("Empty byte array should return true.")
	}
}

func TestUnsupportedAdditionalInformationValues(t *testing.T) {
	notSupportedStartBytes := []byte{0x1C, 0x1D, 0x1E /*AdditionalInfoReserved*/, 0x1F /*AdditionalInfoInfinite*/}

	for _, b := range notSupportedStartBytes {
		ainfo := convertToAdditionalInfo(b)

		shouldPanic(t, func() {
			ainfo.getAdditionalInfoLength()
		})

		if err := Deterministic([]byte{b}); err == nil {
			t.Error("Using AdditionalInfoReserved and AdditionalInfoInfinite should not return true for deterministicy.")
		}
	}
}

func TestGetUnsignedIntegerValue(t *testing.T) {
	for i, testCase := range uint64TestCasesAsBytes {
		res := getUnsignedIntegerValue(testCase, convertToAdditionalInfo(testCase[0]))

		if res != uint64Values[i] {
			t.Errorf("deterministic: getUnsignedIntegerValue, got %v, wanted %v", res, uint64Values[i])
		}
	}
}

func TestAdditionalInfoConversion(t *testing.T) {
	for i := 0; i <= 23; i++ {
		if got := convertToAdditionalInfo(byte(i)); got != AdditionalInfoDirect {
			t.Errorf("deterministic: convertToAdditionalInfo, got %v, wanted %v", got, AdditionalInfoDirect)
		}
	}

	for b, wanted := range map[byte]AdditionalInfo{
		24: AdditionalInfoOneByte,
		25: AdditionalInfoTwoBytes,
		26: AdditionalInfoFourBytes,
		27: AdditionalInfoEightBytes,
		28: AdditionalInfoReserved,
		29: AdditionalInfoReserved,
		30: AdditionalInfoReserved,
		31: AdditionalInfoIndefinite,
	} {
		if got := convertToAdditionalInfo(b); got != wanted {
			t.Errorf("deterministic: convertToAdditionalInfo, got %v, wanted %v", got, wanted)
		}
	}
}

func TestByteStringIsDeterministic(t *testing.T) {
	testBytes := []byte{toCborHeader(TypeBytes, 3), 0x1A, 0x2B, 0x3C}
	testBytesAsCborSequence := multiappend(testBytes, testBytes)

	for _, testCase := range [][]byte{testBytes, testBytesAsCborSequence} {
		if err := Deterministic(testCase); err != nil {
			t.Error("Deterministically encoded byte string should not return false for deterministicy.")
		}
	}
}

func TestLongTextAndByteStringIsDeterministic(t *testing.T) {
	longStr := "olipakerrankilpikonnajakissajotkajuoksivatkilpaajakilpikonnavoitti"
	testBytesAsByteString := multiappend([]byte{toCborHeader(TypeBytes, 24)}, []byte{byte(len(longStr))}, []byte(longStr))
	testBytesAsTextString := multiappend([]byte{toCborHeader(TypeText, 24)}, []byte{byte(len(longStr))}, []byte(longStr))

	for _, testCase := range [][]byte{testBytesAsByteString, testBytesAsTextString} {
		if err := Deterministic(testCase); err != nil {
			t.Error("Deterministically encoded text or byte string should not return false for deterministicy.")
		}
	}
}

func TestTextStringIsDeterministic(t *testing.T) {
	hello := multiappend([]byte{toCborHeader(TypeText, 5)}, []byte("hello"))
	hello2 := multiappend([]byte{toCborHeader(TypeText, 6)}, []byte("hello2"))
	combinedHellos := multiappend(hello, hello2)

	for _, testCase := range [][]byte{hello, hello2, combinedHellos} {
		if err := Deterministic(testCase); err != nil {
			t.Error("Deterministically encoded text string should not return false for deterministicy.")
		}
	}
}

func TestByteStringIsNotDeterministic(t *testing.T) {
	// Claiming it would be followed by 3 bytes but there are only 2.
	testBytes := []byte{toCborHeader(TypeBytes, 3), 0x1A, 0x2B}

	shouldPanic(t, func() {
		textOrByteStringDeterministic(testBytes)
	})
}

func TestArrayIsDeterministic(t *testing.T) {
	arr := multiappend([]byte{toCborHeader(TypeArray, 5)}, uint23, uint24, uint45, uint4294967295, uint18446744073709551615)
	cborSequenceArray := multiappend(arr, arr)

	for _, testCase := range [][]byte{arr, cborSequenceArray} {
		if err := Deterministic(testCase); err != nil {
			t.Error("Deterministically encoded CBOR array should not return false for deterministicy.")
		}
	}
}

func TestEmptyArrayIsDeterministic(t *testing.T) {
	if err := Deterministic([]byte{toCborHeader(TypeArray, 0)}); err != nil {
		t.Error("Empty array should not return false for deterministicy.")
	}
}

func TestLongArrayIsDeterministic(t *testing.T) {
	const numOfItems = 24
	longArr := make([]byte, 2 /*=num of startBytes*/ +len(uint45)*numOfItems)
	longArr[0] = toCborHeader(TypeArray, 24) // ARR (4) + ONE_BYTE (24)
	longArr[1] = byte(numOfItems)

	// Fill in the byte array with 24 times uint45 bytes.
	for i := 2; i < len(longArr); i += len(uint45) {
		for j := 0; j < len(uint45); j++ {
			longArr[i+j] = uint45[j]
		}
	}

	cborSequenceArray := multiappend(longArr, longArr)

	for _, testCase := range [][]byte{longArr, cborSequenceArray} {
		if err := Deterministic(testCase); err != nil {
			t.Error("Deterministically encoded CBOR array should not return false for deterministicy.")
		}
	}
}

func TestArraysNumberOfItemsIsWrong(t *testing.T) {
	// There is one item too little. The additional information claims that there are 5 but there are only 4.
	arr := multiappend([]byte{toCborHeader(TypeArray, 5)}, uint23, uint24, uint45, uint4294967295)
	shouldPanic(t, func() {
		Deterministic(arr)
	})
}

func TestArrayContainsNonDeterministicCbor(t *testing.T) {
	arr := multiappend([]byte{toCborHeader(TypeArray, 5)}, uint23, uint24, uint45, convertToNonDeterministicUintHelper(uint255, 0x19), uint4294967295)
	if err := Deterministic(arr); err == nil {
		t.Error("Deterministically encoded CBOR array should not return true for deterministicy.")
	}
}

func TestMapIsDeterministic(t *testing.T) {
	hello := multiappend([]byte{toCborHeader(TypeBytes, 5)}, []byte("hello"))
	hello2 := multiappend([]byte{toCborHeader(TypeBytes, 6)}, []byte("hello2"))
	testMap := multiappend([]byte{toCborHeader(TypeMap, 2)}, hello, uint24, hello2, uint4294967295)
	cborSequenceMap := multiappend(testMap, testMap)

	for _, testCase := range [][]byte{testMap, cborSequenceMap} {
		if err := Deterministic(testCase); err != nil {
			t.Error("Deterministically encoded CBOR map should not return false for deterministicy.")
		}
	}
}

func TestMapWithDuplicateKeysIsNotDeterministic(t *testing.T) {
	hello := multiappend([]byte{toCborHeader(TypeBytes, 5)}, []byte("hello"))
	testMap := multiappend([]byte{toCborHeader(TypeMap, 2)}, hello, uint24, hello, uint4294967295)

	if err := Deterministic(testMap); err == nil {
		t.Error("CBOR map with duplicate keys should not return true for deterministicy.")
	}
}

func TestMapWithUnorderedKeysIsNotDeterministic(t *testing.T) {
	hello := multiappend([]byte{toCborHeader(TypeBytes, 5)}, []byte("hello"))
	hello2 := multiappend([]byte{toCborHeader(TypeBytes, 6)}, []byte("hello2"))

	// 0b01100101 (0x65) should become before 0b01100110 (0x66), so this doesn't follow the lexicographical order.
	testMap := multiappend([]byte{toCborHeader(TypeMap, 2)}, hello2, uint24, hello, uint4294967295)

	if err := Deterministic(testMap); err == nil {
		t.Error("CBOR map with unordered keys should not return true for deterministicy.")
	}
}

func TestMapWithDuplicateNotSequentialKeysIsNotDeterministic(t *testing.T) {
	hello := multiappend([]byte{toCborHeader(TypeBytes, 5)}, []byte("hello"))
	hello2 := multiappend([]byte{toCborHeader(TypeBytes, 6)}, []byte("hello2"))
	testMap := multiappend([]byte{toCborHeader(TypeMap, 3)}, hello, uint24, hello2, uint4294967295, hello, uint18446744073709551615)

	if err := Deterministic(testMap); err == nil {
		t.Error("CBOR map with duplicate keys (which are not following each other) should not return true for deterministicy.")
	}
}

func TestMapsNumberOfItemsIsWrong(t *testing.T) {
	// There is one item too little. The additional information claims that there are
	// 2 pairs (=4 items) but there are only 1 pair and 1 item (3 items).
	testMap := multiappend([]byte{toCborHeader(TypeMap, 2)}, uint23, uint24, uint45)
	shouldPanic(t, func() {
		Deterministic(testMap)
	})
}

// The test web bundle have generated only contains deterministic CBOR. This test is to make
// sure that we don't accidentally break our deterministic checker. It contains a lot of
// nested CBOR types so it might catch something that has not been covered with other tests.
func TestWebBundleHasDeterministicCbor(t *testing.T) {
	bundleFile, err := os.Open("../../integrityblock/testfile.wbn")
	if err != nil {
		t.Error("Failed to open the test file")
	}
	defer bundleFile.Close()
	bundleBytes, err := ioutil.ReadAll(bundleFile)
	if err != nil {
		t.Error("Failed to read the test file")
	}

	if err := Deterministic(bundleBytes); err != nil {
		t.Error("testfile.wbn only contains deterministic CBOR and this shouldn't fail.")
	}
}

// Helper functions:

func convertToNonDeterministicUintHelper(deterministicUintBytes []byte, firstByte byte) []byte {
	newLength := convertToAdditionalInfo(firstByte).getAdditionalInfoLength()
	shift := 0
	if convertToAdditionalInfo(deterministicUintBytes[0]) != AdditionalInfoDirect {
		// If it's a direct value, the first byte is not part of the actual value.
		shift = 1
	}

	emptyExtraBytes := make([]byte, newLength-len(deterministicUintBytes)+shift)

	return multiappend([]byte{firstByte}, emptyExtraBytes, deterministicUintBytes[shift:])
}

func multiappend(inputs ...[]byte) []byte {
	var res []byte
	for _, input := range inputs {
		res = append(res, input...)
	}
	return res
}

func shouldPanic(t *testing.T, f func()) {
	t.Helper()
	defer func() { _ = recover() }()
	f()
	t.Errorf("should have panicked")
}

func toCborHeader(majorType Type, ainfoValue int) byte {
	if ainfoValue < 0 || ainfoValue > 27 {
		panic("The value of the additional information should be between 0 and 27 inclusive.")
	}
	return byte(majorType) | byte(ainfoValue)
}

File: js/sign/tests/deterministic_test.js
import * as det from '../lib/cbor/deterministic.js';
import * as ainfo from '../lib/cbor/additionalinfo.js';

const uInts = {
  0: new Uint8Array([0x00]),
  10: new Uint8Array([0x0a]),
  23: new Uint8Array([0x17]),
  24: new Uint8Array([0x18, 0x18]),
  45: new Uint8Array([0x18, 0x2d]),
  255: new Uint8Array([0x18, 0xff]),
  256: new Uint8Array([0x19, 0x01, 0x00]),
  5000: new Uint8Array([0x19, 0x13, 0x88]),
  65535: new Uint8Array([0x19, 0xff, 0xff]),
  65536: new Uint8Array([0x1a, 0x00, 0x01, 0x00, 0x00]),
  4294967: new Uint8Array([0x1a, 0x00, 0x41, 0x89, 0x37]),
  4294967295: new Uint8Array([0x1a, 0xff, 0xff, 0xff, 0xff]),
  // This is also the max length of Uint8Array on node.
  4294967296: new Uint8Array([
    0x1b, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x00,
  ]),
  // Number.MAX_SAFE_INTEGER
  9007199254740991: new Uint8Array([
    0x1b, 0x00, 0x1f, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  ]),
  '9223372036854775807': new Uint8Array([
    0x1b, 0x7f, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  ]),
  // Max CBOR supported value 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF 0xFF.
  '18446744073709551615': new Uint8Array([
    0x1b, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
  ]),
};

describe('Deterministic check - Unsigned Integers', () => {
  it('works for positive integers, both single and sequences.', () => {
    for (const testCase of Object.values(uInts)) {
      det.checkDeterministic(testCase);
    }
    // CBOR sequence of positive integers
    det.checkDeterministic(Buffer.concat(Object.values(uInts)));
  });

  it("doesn't throw error for empty byte array.", () => {
    det.checkDeterministic(new Uint8Array());
  });

  it('calculates the value of an unsigned integer correctly.', () => {
    for (const testCase of Object.entries(uInts)) {
      const [expectedValue, valueInBytes] = testCase;
      const calculatedValue = det.getUnsignedIntegerValue(
        valueInBytes,
        ainfo.convertToAdditionalInfo(valueInBytes[0])
      );
      expect(calculatedValue).toBe(BigInt(expectedValue));
    }
  });

  // If the value of the unsigned integer is kept the same but there are
  // additional non-necessary empty bytes added in-front of the bytes of the
  // actual value, it's no longer considered as deterministically encoded CBOR
  // and should throw an error.
  it('fails for non-deterministically encoded positive integers.', () => {
    // The additional information values for representing a number with
    // 1, 2, 4 or 8 bytes.
    const firstBytes = [0x18, 0x19, 0x1a, 0x1b];

    for (const testCase of Object.values(uInts)) {
      for (const firstByte of firstBytes) {
        const ai = ainfo.convertToAdditionalInfo(firstByte);
        const newLength = ainfo.getAdditionalInfoLength(ai);

        // Cannot represent too big number with too little amount of bytes.
        if (testCase.length > newLength) {
          continue;
        }

        const nonDeterministicBytes = convertToNonDeterministicUintHelper(
          testCase,
          firstByte
        );
        const valueOfOriginalByteArr = det.getUnsignedIntegerValue(
          testCase,
          ainfo.convertToAdditionalInfo(testCase[0])
        );
        const valueOfNonDeterministicByteArr = det.getUnsignedIntegerValue(
          nonDeterministicBytes,
          ainfo.convertToAdditionalInfo(nonDeterministicBytes[0])
        );
        if (valueOfOriginalByteArr !== valueOfNonDeterministicByteArr) {
          throw new Error(
            'Values of original bytes and non-deterministic bytes should match.'
          );
        }

        expect(() =>
          det.checkDeterministic(nonDeterministicBytes)
        ).toThrowError();
      }
    }
  });
});

describe('Deterministic check - Additional information', () => {
  it('converts additional info correctly.', () => {
    for (let i = 0; i <= 23; i++) {
      expect(ainfo.convertToAdditionalInfo(i)).toEqual(
        ainfo.AdditionalInfo.Direct
      );
    }

    for (const entry of Object.entries({
      24: ainfo.AdditionalInfo.OneByte,
      25: ainfo.AdditionalInfo.TwoBytes,
      26: ainfo.AdditionalInfo.FourBytes,
      27: ainfo.AdditionalInfo.EightBytes,
    })) {
      const [key, value] = entry;
      expect(ainfo.convertToAdditionalInfo(key)).toEqual(value);
    }

    for (const notSupportedAdditionalInfo of [28, 29, 30, 31]) {
      expect(() =>
        ainfo.convertToAdditionalInfo(notSupportedAdditionalInfo)
      ).toThrowError();
    }
  });
});

describe('Deterministic check - ByteString and Text', () => {
  it('works for byte strings.', () => {
    const testBytes = new Uint8Array([0x43, 0x1a, 0x2b, 0x3c]);
    det.checkDeterministic(testBytes);
    det.checkDeterministic(new Uint8Array([...testBytes, ...testBytes]));
  });

  it('works for text.', () => {
    const testBytes = new Uint8Array([0x63, 0x1a, 0x2b, 0x3c]);
    det.checkDeterministic(testBytes);
    det.checkDeterministic(new Uint8Array([...testBytes, ...testBytes]));
  });

  it('works for long text and byte strings.', () => {
    const longStr =
      'olipakerrankilpikonnajakissajotkajuoksivatkilpaajakilpikonnavoitti';

    const testBytesAsByteString = new Uint8Array([
      0x58,
      longStr.length,
      ...new Uint8Array(Buffer.from(longStr)),
    ]);

    const testBytesAsTextString = new Uint8Array([
      0x78,
      longStr.length,
      ...new Uint8Array(Buffer.from(longStr)),
    ]);

    det.checkDeterministic(testBytesAsByteString);
    det.checkDeterministic(testBytesAsTextString);
    det.checkDeterministic(
      new Uint8Array([...testBytesAsByteString, ...testBytesAsByteString])
    );
    det.checkDeterministic(
      new Uint8Array([...testBytesAsTextString, ...testBytesAsTextString])
    );
  });

  it('detects non-deterministic byte string and text.', () => {
    // Claiming they would be followed by 3 bytes but there are either 2 or 4.
    const b01000011 = 0x43;
    const b01100011 = 0x63;
    const twoRandomBytes = [0x1a, 0x2b];
    const fourRandomBytes = [0x1a, 0x2b, 0x1a, 0x2b];

    expect(() =>
      det.checkDeterministic(new Uint8Array([b01000011, ...twoRandomBytes]))
    ).toThrowError();
    expect(() =>
      det.checkDeterministic(new Uint8Array([b01000011, ...fourRandomBytes]))
    ).toThrowError();

    expect(() =>
      det.checkDeterministic(new Uint8Array([b01100011, ...twoRandomBytes]))
    ).toThrowError();
    expect(() =>
      det.checkDeterministic(new Uint8Array([b01100011, ...fourRandomBytes]))
    ).toThrowError();
  });
});

// Helper functions.

function convertToNonDeterministicUintHelper(
  deterministicUintBytes,
  firstByte
) {
  const newLength = ainfo.getAdditionalInfoLength(
    ainfo.convertToAdditionalInfo(firstByte)
  );

  const ainfoIsDirect =
    ainfo.convertToAdditionalInfo(deterministicUintBytes[0]) ===
    ainfo.AdditionalInfo.Direct;

  // Defines if the first byte is part of the actual value or not.
  const shift = ainfoIsDirect ? 0 : 1;

  const emptyExtraBytes = new Uint8Array(
    newLength - deterministicUintBytes.length + shift
  );

  return new Uint8Array([
    firstByte,
    ...emptyExtraBytes,
    ...deterministicUintBytes.slice(shift),
  ]);
}
